### ISSUE TITLE
Large playlist = "Error in script init(): CPU limit"

### ISSUE THREAD

**Issue by JyeSmith** [GilDev/TaraniTunes#7]
Im only able to get 35 songs in my playlist before "Error in script init(): CPU limit".  Is it just me or do you get the same?

**Comment by impas1k** [comment]
The same issue has only 13 songs. Can even launch the script. Error in script init(): CPU limit

**Comment by GilDev** [comment]
I don't know about the maximum number of songs you can use, but there shouldn't be a limit that low.
**Be sure** that there is no newline at the end of the file, some editors like “nano” automatically add one even if you delete it. Also any error in the “playlist.txt” file will result in a “CPU limit” error, maybe non-ASCII characters could trigger that.
I really need to enhance the error handling on the playlist file, but I don't have time right now… =/

**Comment by impas1k** [comment]
Just delete 10 songs, the script is running now. Will try to bind the switches now ) thanks bro

**Comment by GilDev** [comment]
Yeah, really make sure “playlist.txt” is correctly formatted, that's the deal breaker right now…

**Comment by JyeSmith** [comment]
@impas1k I played around with the script a bit more last night and this is where Im at.  Add the playlist to the lua script and it will run fine with a large number of songs.  I couldnt find out if the issue was memory, cpu, or the script taking too long to execute and read the playlist.txt file.

Heres my current branch with a 'hard coded' playlist along with more changes;

https://github.com/JyeSmith/TaraniTunes/blob/JyeSmith-TaraniTunes/iTunes.lua

```
playlist = 
{{"", ""}, {"", ""},
{"Music 1", "Music1"},
{"Music 2", "Music2"},
{"Music 3", "Music3"},
{"", ""}, {"", ""}}
```

and delete

```
	-- Build playlist
	playlist = {{"", ""}, {"", ""}}
	playlistFile = io.open(playlistFilename, "r")

	if (playlistFile == nil) then
		error({"\"" .. playlistFilename .. "\"", "file not found, please", "read the \"README.md\"", "file for installation", "instructions."})
		return
	end
	
	repeat
		songName = ""
		songFilename = ""

		-- Determine song name
		char = ""
		repeat
			songName = songName .. char
			char = io.read(playlistFile, 1)
		until char == ":"

		-- Determine song filename
		char = ""
		repeat
			songFilename = songFilename .. char
			char = io.read(playlistFile, 1)
		until char == "\n" or char == "\r" or char == ""
		if char == "\r" then io.read(playlistFile, 1) end -- Discard newline character for Windows ending

		playlist[#playlist + 1] = {songName, songFilename} -- table.insert() doesn't work

	until char == ""

	io.close(playlistFile)

	-- These blank starting and ending values are used when selecting top or bottom songs
	playlist[#playlist + 1] = {"", ""}
	playlist[#playlist + 1] = {"", ""}
```

**Comment by GilDev** [comment]
That's great if it works, but of course that cant be a permanent solution, it's just not convenient. =)
I'm sure there is some way to make the program read a playlist with plenty of songs.
My current playlist have 37 songs and I have no problem, here it is:
```
Alan Walker - Faded:MAlan1
Alan Walker - Force:MAlan2
Alan Walker - The Spectre:MAlan3
Andrew Applepie - Secrets:MAndr1
BUNT. - Young Hearts:MBUNT1
Cascada - Everytime We Touch:MCasc1
Cazzette - Blind Heart:MCazz1
Cazzette - She Wants Me Dead:MCazz2
Cazzette - Together:MCazz3
Debut de Soiree - Nuit de Folie:MDebu1
Deorro - Five Hours:MDeor1
Dooqu - Oh Yeah:MDooq1
EDX - Feel The Rush:MEDX1
Ilona - Un Monde Parfait:MIlon1
Jamie Berry - Delight:MJami1
Jamie Berry - Kiss Me:MJami2
Jamie Berry - Old Records:MJami3
Jamie Berry - Peeping Tom:MJami4
Jamie Berry - Sour Puss:MJami5
Jamie Berry - Twitch:MJami6
Les Ratz - Pas de Panique:MLesR1
Lipps - Funkytown:MLipp1
MitchiriNeko - Cat March:MMitc1
Not Your Dope - 5 A.M.:MNotY1
Nyan Cat:MNyan1
TheFatRat - Fly Away:MTheF1
TheFatRat - Jackpot:MTheF2
TheFatRat - Monody:MTheF3
TheFatRat - Oblivion:MTheF4
TheFatRat - The Calling:MTheF5
TheFatRat - Unity:MTheF6
TheFatRat - Xenogenesis:MTheF7
Vexento - Lonely Dance:MVexe1
Vexento - Masked Heroes:MVexe2
Vexento - Pixel Party:MVexe3
Vexento - Tevo:MVexe4
Vexento - Thirsty:MVexe5
```

I made it with Sublime Text 3, making sure there was no newline at the end, that no filename has more than 6 characters and that only ASCII characters are used.

**Comment by JyeSmith** [comment]
@GilDev Can you try cut@paste your playlist a couple of times to make it longer and see what happens.

And its no hard to edit a lua script or playlist, they are both simple text files.  I auto generate my playlist using the below.
```

newPlayFile="luaPlayList_$(hexdump -e '/1 "%02x"' -n3 < /dev/urandom).txt"

fileExtension=".wav"

files=*.mp3

for f in $files
do
  echo $f
  fileLength=$(ffmpeg -i "$f" 2>&1 | grep "Duration" | cut -d ' ' -f 4 | sed s/,// | sed 's@\..*@@g' | awk '{ split($1, A, ":"); split(A[3], B, "."); print 3600*A[1] + 60*A[2] + B[1] }')
  echo $fileLength
  newFileName="0$(cat /dev/urandom | tr -cd 'a-f0-9' | head -c 5)"
  shortFileName=$(echo $f | cut -c6-30)
  echo "{\"${shortFileName}\", \"$newFileName\", \"$fileLength\"},">>$newPlayFile
  ffmpeg -i "$f" -acodec pcm_s16le -ac 1 -ar 16000 $newFileName$fileExtension
done

```

**Comment by impas1k** [comment]
So 

playlist = 
{{"", ""}, {"", ""},
{"Music 1", "Music1"},
{"Music 2", "Music2"},
{"Music 3", "Music3"},
{"", ""}, {"", ""}}

and 

newPlayFile="luaPlayList_$(hexdump -e '/1 "%02x"' -n3 < /dev/urandom).txt"

fileExtension=".wav"

files=*.mp3

for f in $files
do
  echo $f
  fileLength=$(ffmpeg -i "$f" 2>&1 | grep "Duration" | cut -d ' ' -f 4 | sed s/,// | sed 's@\..*@@g' | awk '{ split($1, A, ":"); split(A[3], B, "."); print 3600*A[1] + 60*A[2] + B[1] }')
  echo $fileLength
  newFileName="0$(cat /dev/urandom | tr -cd 'a-f0-9' | head -c 5)"
  shortFileName=$(echo $f | cut -c6-30)
  echo "{\"${shortFileName}\", \"$newFileName\", \"$fileLength\"},">>$newPlayFile
  ffmpeg -i "$f" -acodec pcm_s16le -ac 1 -ar 16000 $newFileName$fileExtension
done
 

need to be added in LUA and 

-- Build playlist
	playlist = {{"", ""}, {"", ""}}
	playlistFile = io.open(playlistFilename, "r")

	if (playlistFile == nil) then
		error({"\"" .. playlistFilename .. "\"", "file not found, please", "read the \"README.md\"", "file for installation", "instructions."})
		return
	end
	
	repeat
		songName = ""
		songFilename = ""

		-- Determine song name
		char = ""
		repeat
			songName = songName .. char
			char = io.read(playlistFile, 1)
		until char == ":"

		-- Determine song filename
		char = ""
		repeat
			songFilename = songFilename .. char
			char = io.read(playlistFile, 1)
		until char == "\n" or char == "\r" or char == ""
		if char == "\r" then io.read(playlistFile, 1) end -- Discard newline character for Windows ending

		playlist[#playlist + 1] = {songName, songFilename} -- table.insert() doesn't work

	until char == ""

	io.close(playlistFile)

	-- These blank starting and ending values are used when selecting top or bottom songs
	playlist[#playlist + 1] = {"", ""}
	playlist[#playlist + 1] = {"", ""}

deleted?

**Comment by impas1k** [comment]
maybe you can update LUA?

**Comment by JyeSmith** [comment]
@impas1k try this

https://github.com/JyeSmith/TaraniTunes/blob/for_impas1k/iTunes.lua

Add your playlist at the top of the file and then save it in the Telemetry folder as normal.

**Comment by GilDev** [comment]
@JyeSmith Sure but to have the best UX, an external “playlist.txt” file with no code to possibly alter seems better to me, even if most Taranis users know a bit about code manipulation.
If that causes memory problems, yeah, a Lua playlist could solve the problem, but that shouldn't be an issue anyway by reading the playlist file line by line.
I tried copying and pasting my entire playlist once more and indeed I get a “CPU limit” error. Further testing is needed to fix this.

**Comment by JyeSmith** [comment]
Agreed.

The solution may be to separate the playlist into another lua located in the SOUNDS dir and use `dofile `or `loadfile `.

**Comment by GilDev** [comment]
Yeah that should work, but I still don't know why it would work like this and not with a parsed text file. Memory shouldn't be the issue if the playlist is read line by line and stored in an array identical to the hardcoded one you made.

**Comment by JyeSmith** [comment]
Do you know if there is a time limit for execution of init()?  Could it be taking too long to read the txt file?

**Comment by impas1k** [comment]
@JyeSmith flawlessly! Thanks a lot! Following the updates )

**Comment by GilDev** [comment]
Fixed (I hope so) in 4f39d45.
